packages/core/strapi/src/cli/commands/generate-schema.ts

```diff
@@ -0,0 +1,210 @@
+const restrictToApiContentTypes = (uid: string): boolean => {
+  try {
+    return uid.startsWith('api::');
+  } catch (e) {
+    console.error('Error checking UID:', e);
+    return false;
+  }
+};
+
+const allCrudOperationNames = ['create', 'update', 'delete', 'findOne', 'find'];
+
+const crudOperationNames = {
+  collectionType: allCrudOperationNames,
+  singleType: allCrudOperationNames.filter((op) => op !== 'find'),
+};
+
+const action = async ({ silent, outDir }: CmdOptions) => {
+  if (silent) {
+    console.log = () => {};
+  }
+
+  console.log('Generating schema.ts...');
+  const outputDir = outDir ?? 'types';
+
+  // Create and register a Strapi instance
+  const appContext = await compileStrapi();
+  const app = await createStrapi(appContext).register();
+
+  // Generate types with modifications. The Modifications are:
+  // - Filter out content types that are not API content types
+  // - Filter out attributes that are private or password
+  await tsUtils.generators.generate({
+    strapi: app,
+    pwd: appContext.appDir,
+    rootDir: outputDir,
+    logger: {
+      silent,
+    },
+    artifacts: {
+      contentTypes: true,
+      components: true,
+    },
+    filters: {
+      contentTypes(contentType: Struct.CollectionTypeSchema) {
+        return restrictToApiContentTypes(contentType.uid);
+      },
+    },
+    transforms: {
+      contentTypes(contentType: Struct.CollectionTypeSchema) {
+        // Strip private attributes
+        const attributes = Object.fromEntries(
+          Object.entries(contentType.attributes).filter(
+            ([, attr]) =>
+              attr.type !== 'password' &&
+              // @ts-expect-error TODO: fix this
+              !attr.private
+          )
+        );
+        return { ...contentType, attributes };
+      },
+    },
+  });
+
+  const contentTypes = app.get('content-types').getAll();
+  const apiContentTypes: [string, Struct.CollectionTypeSchema | Struct.SingleTypeSchema][] =
+    Object.entries(contentTypes).filter(([uid]) => restrictToApiContentTypes(uid)) as [
+      string,
+      Struct.CollectionTypeSchema | Struct.SingleTypeSchema,
+    ][];
+
+  // Map UIDs to generated type names
+  const uidToGeneratedTypeName = new Map<string, string>();
+  const allTypeNames = apiContentTypes.map(([uid, contentType]) => {
+    const {
+      info: { singularName },
+    } = contentType;
+
+    const formattedSingularName = singularName
+      .split('-')
+      .map((part) => part.charAt(0).toUpperCase() + part.slice(1))
+      .join('');
+
+    const typeName = `Api${formattedSingularName}${formattedSingularName}`;
+    uidToGeneratedTypeName.set(uid, typeName);
+
+    return typeName;
+  });
+
+  // Generate a single import statement for generated types
+  const combinedImportStatement = `import { ${allTypeNames.join(', ')} } from './generated/contentTypes';`;
+
+  // Helper function to generate TypeScript interface definitions
+  const generateInterfaceDefinition = (
```

Comment:
I believe it could be easier to use the TS factory API to generate the interfaces rather than trying to generate raw strings, what do you think?

Reply:
Also, maybe we could consider moving the schema generation part to the typescript generators? Or do we want to keep the logic in the action on purpose?